File: cardPickerPolicy.js
import * as tf from '@tensorflow/tfjs';
import { config, print, debug } from './config.js';
import { BasePolicy } from './policies/basePolicy.js';
import { BaseRLPolicy } from './policies/baseRLPolicy.js';

export class PolicyFactory{    
  static makePolicy = function(policyType, encoder, learningRate) {    
    switch (policyType) {
      case POLICY_TYPES.MANUAL_UPDATES:
        return new ManualUpdatesPolicy(encoder, learningRate)
      case POLICY_TYPES.RL_PICK_MULTIPLE_CARDS:
        return new HandPermutationDensePolicy(encoder, null, config.HIDDEN_LAYERS, learningRate);
      case POLICY_TYPES.RL_PICK_CARD_AT_A_TIME:
        var policy = new PickCardAtATimeDensePolicy(encoder, null, config.HIDDEN_LAYERS, learningRate);
        return policy;
      default:
          throw 'Invalid policy type'
    }
  }
}

export const POLICY_TYPES = {
  RL_PICK_CARD_AT_A_TIME: 'RL_PICK_CARD_AT_A_TIME',
  RL_PICK_MULTIPLE_CARDS: 'RL_PICK_MULTIPLE_CARDS',
  MANUAL_UPDATES: 'MANUAL_UPDATES'
}

class ManualUpdatesPolicy extends BasePolicy{
  constructor(encoder) {
    super()
    this.encoder = encoder
    this.learningRate = 0.1;
    var lenUniqueCards = config.MAX_HERO_ALLY_ATTACK - config.MIN_HERO_ALLY_ATTACK + 1
    this.policyTensor = tf.randomUniform([1, lenUniqueCards], 0, 1)
  }

  onMatchStart = function() {
    this.cardIndecesPicked = []
  }

  update = function (score) {
      var avgScore = score / this.cardIndecesPicked.length
      this.cardIndecesPicked.forEach((index) => {
          var oneHotCard = this.encoder.getOneHotEncodedCardByIndex(index)
          var updates = oneHotCard.map(n => n *= avgScore)
          this.updatePolicyMatrix(tf.tensor1d(updates, 'float32'))
        })
  }

  updatePolicyMatrix = function(policyTensor) {
    policyTensor = policyTensor.mul(this.learningRate)
    this.policyTensor = this.policyTensor.add(policyTensor)
  }

  pickCard = function(hand) {
    //run hand through policy network to find best one
    //turn hand to matrix then tensor
    var oneHotHand = this.encoder.oneHotEncodeCards(hand)
    var handTensor = tf.tensor(oneHotHand, [oneHotHand.length, oneHotHand[0].length])
    var result = handTensor.matMul(this.policyTensor, false, true)
    var maxIndexInHand = this.pickIndexOverDistribution(result)
    var indexInUniqueCards = this.encoder.getIndexOfCard(hand[maxIndexInHand])
    this.cardIndecesPicked.push(indexInUniqueCards)
    var card = hand[maxIndexInHand]
    return card
  }

  pickCards = function(hand) {
    var cards = [this.pickCard(hand)]
    return cards
  }

  peek = function(){
    return this.policyTensor.arraySync()
  }
}

export class PickCardAtATimeDensePolicy extends BaseRLPolicy{
  constructor(encoder, predictionPicker, hiddenLayers = 0, learningRate = 0.1) {
    super(encoder, predictionPicker, learningRate);
    this.inputUnits = this.encoder.inputUnits;
    this.hiddenLayers = hiddenLayers;
    this.predictionPicker = predictionPicker;
    this.model = tf.sequential();
    this.model.initialWeights = tf.randomUniform([this.inputUnits, 1], 0, 1)
    this.model.add(tf.layers.dense({units: config.HIDDEN_UNITS, inputShape: [this.inputUnits], useBias: true}));
    this.model.add(tf.layers.dense({units: 1, useBias: true}));    
    this.model.compile({loss: 'meanSquaredError', optimizer: tf.train.adam(this.learningRate)});  
  }

  update = async function (score) {
    var avgScore = score / this.turns;
    var inputMatrix = tf.tensor2d([...this.pickedActions], [this.pickedActions.length, this.inputUnits])
    var labels = tf.tensor1d([...this.pickedActions.map(n => avgScore)], 'float32')
    await this.updateModel(inputMatrix, labels)
  }

  getUniqueCards = function(hand) {
    var uniqueCards = []
    hand.forEach((card) => {
      if(uniqueCards.find(n => n.name == card.name) == undefined) {
        uniqueCards.push(card)
      }
    })
    return uniqueCards
  }

  encodedCardToColumnTensor = function(encodedCard, inputUnits) {
    var combinationTensor = tf.tensor(encodedCard,[inputUnits,1], 'float32') 
    return combinationTensor
  }

  pickCards = function(hand, world) {
    // encode cards into tensors
    var uniqueCards = this.getUniqueCards(hand);
    var inputTensors = uniqueCards.map(n => {
      var encoded = this.encoder.encodeCard(n, world)
      return this.encodedCardToColumnTensor(encoded, this.inputUnits)
    });
    // Build matrix of permutations
    var inputMatrix = inputTensors.reduce((a, b) => a.concat(b,1)).transpose()
    // Run each card through model to get probability of picking each card
    var predictions = this.model.predict(inputMatrix)
    var testPred = predictions.arraySync()
    var testPeek = this.peek();
    var pickedCardIndex = this.predictionPicker(predictions)
    var pickedCard = uniqueCards[pickedCardIndex]
    // Save picked Card as index array
    this.pickedActions.push(inputMatrix.arraySync()[pickedCardIndex])
    print({damage:pickedCard.attack, cost:pickedCard.cost})
    // var newHistory = pickedCard.attack
    // this.history.push(newHistory)
    this.turns++;
    return [pickedCard]
  }
}

export class HandPermutationDensePolicy extends BaseRLPolicy{
  constructor(encoder, predictionPicker, hiddenLayers=0, learningRate = 0.1) {
    super(encoder, predictionPicker, learningRate);
    this.inputUnits = this.lenUniqueCards*config.HERO_BUDGET;
    this.model = tf.sequential();
    this.model.initialWeights = tf.randomUniform([this.inputUnits, 1], 0, 1)
    this.model.add(tf.layers.dense({units: config.HIDDEN_UNITS, inputShape: [this.inputUnits], useBias: true}));
    // this.model.add(tf.layers.layerNormalization({axis: -1}));
    for(var i = 0; i < hiddenLayers; i++) {
      this.model.add(tf.layers.dense({units: config.HIDDEN_UNITS, inputShape: [config.HIDDEN_UNITS], useBias: true}));
    }
    this.model.add(tf.layers.dense({units: 1, useBias: true}));    
    this.model.compile({loss: 'meanSquaredError', optimizer: tf.train.adam(this.learningRate)});   
    this.oneHotCombinationToColumnTensor = this.oneHotCombinationToColumnTensor.bind(this)
  }
  
  update = async function (score) {
    var avgScore = score / this.turns;
    var inputMatrix = tf.tensor2d([...this.pickedActions], [this.pickedActions.length, this.inputUnits])
    var labels = tf.tensor1d([...this.pickedActions.map(n => avgScore)], 'float32')
    await this.updateModel(inputMatrix, labels)
  }

  oneHotCombinationToColumnTensor = function(oneHotCombination, inputUnits) {
    var combinationTensor = tf.tensor(oneHotCombination,[oneHotCombination.length,this.lenUniqueCards], 'float32')
    var paddedCombinationTensor = this.padCombinationTensor(combinationTensor)
    // reshape combinationTensort to be 1 column
    var reshapedCombination = paddedCombinationTensor.reshape([inputUnits, 1])      
    return reshapedCombination
  }

  getProbabilityForPermutation = function(permutation, allPermutations, probabilities){
    for(var i = 0; i < allPermutations.length; i++){
      if(permutation.toString() === allPermutations[i].toString()){
        return probabilities[i]
      }
    }
    throw new Error('Permutation not found')
  }

  pickCards = function(hand) {
    // Make all combinations of up to n cards from hand
    var possiblePlayedCards = this.buildNCardPermutations(config.HERO_BUDGET, hand)
    var inputTensors = possiblePlayedCards.map(n => {
      var oneHotCombination = this.encoder.oneHotEncodeCards(n)
      return this.oneHotCombinationToColumnTensor(oneHotCombination, this.inputUnits)
    })
    // Build matrix of permutations
    var inputMatrix = inputTensors.reduce((a, b) => a.concat(b,1)).transpose()
    // Run each combination through model to get probability of picking each card
    var predictions = this.model.predict(inputMatrix)
    var pickedCombinationIndex = this.predictionPicker(predictions)
    var pickedCombination = possiblePlayedCards[pickedCombinationIndex]
    // var test = inputMatrix.arraySync()
    // Save actions for update
    this.pickedActions.push(inputMatrix.arraySync()[pickedCombinationIndex])
    print(pickedCombination.map(n => ({damage:n.attack, cost:n.cost})))
    var newHistory = pickedCombination.map(n => n.attack)
    this.history.push(newHistory)
    this.turns++;
    return pickedCombination
  }
}

File: python/cardPickerPolicy.py
# import * as tf from '@tensorflow/tfjs';
# import { config, print, debug } from './config.js';
# import { BasePolicy } from './policies/basePolicy.js';
# import { BaseRLPolicy } from './policies/baseRLPolicy.js';
# 
# export class PolicyFactory{    
#   static makePolicy = function(policyType, encoder, learningRate) {    
#     switch (policyType) {
#       case POLICY_TYPES.MANUAL_UPDATES:
#         return new ManualUpdatesPolicy(encoder, learningRate)
#       case POLICY_TYPES.RL_PICK_MULTIPLE_CARDS:
#         return new HandPermutationDensePolicy(encoder, null, config.HIDDEN_LAYERS, learningRate);
#       case POLICY_TYPES.RL_PICK_CARD_AT_A_TIME:
#         var policy = new PickCardAtATimeDensePolicy(encoder, null, config.HIDDEN_LAYERS, learningRate);
#         return policy;
#       default:
#           throw 'Invalid policy type'
#     }
#   }
# }


# the following is the above commented javascript code converted to python

import tensorflow as tf
import numpy as np
import random
import config
from policies.baseRLPolicy import BaseRLPolicy
from policies.basePolicy import BasePolicy

POLICY_TYPES = {
    RL_PICK_CARD_AT_A_TIME: 'RL_PICK_CARD_AT_A_TIME',
    RL_PICK_MULTIPLE_CARDS: 'RL_PICK_MULTIPLE_CARDS',
    MANUAL_UPDATES: 'MANUAL_UPDATES'
}

class PolicyFactory:
    def makePolicy(policyType, encoder, learningRate):
        switcher = {
            POLICY_TYPES.MANUAL_UPDATES: ManualUpdatesPolicy(encoder, learningRate),
            POLICY_TYPES.RL_PICK_MULTIPLE_CARDS: HandPermutationDensePolicy(encoder, None, config.HIDDEN_LAYERS, learningRate),
            POLICY_TYPES.RL_PICK_CARD_AT_A_TIME: PickCardAtATimeDensePolicy(encoder, None, config.HIDDEN_LAYERS, learningRate)
        }
        return switcher.get(policyType, 'Invalid policy type')

class ManualUpdatesPolicy(BasePolicy):
    def __init__(self, encoder, predictionPicker, learningRate = 0.1):
        super().__init__(encoder, predictionPicker, learningRate)
        self.inputUnits = self.lenUniqueCards
        self.model = tf.keras.Sequential()
        self.model.add(tf.keras.layers.Dense(config.HIDDEN_UNITS, input_shape=(self.inputUnits,), use_bias=True))
        self.model.add(tf.keras.layers.Dense(1, use_bias=True))
        self.model.compile(loss='meanSquaredError', optimizer=tf.keras.optimizers.Adam(self.learningRate))

    def update(self, score):
        avgScore = score / self.turns
        inputMatrix = tf.convert_to_tensor(self.pickedActions, dtype=tf.float32)
        labels = tf.convert_to_tensor([avgScore] * len(self.pickedActions), dtype=tf.float32)
        self.updateModel(inputMatrix, labels)

    def encodedCardToColumnTensor(self, encodedCard, inputUnits):
        combinationTensor = tf.reshape(tf.convert_to_tensor(encodedCard, dtype=tf.float32), [inputUnits, 1])
        return combinationTensor

    def pickCards(self, hand, world):
        # encode cards into tensors
        uniqueCards = self.getUniqueCards(hand)
        inputTensors = [self.encodedCardToColumnTensor(self.encoder.encodeCard(n, world), self.inputUnits) for n in uniqueCards]
        # Build matrix of permutations
        inputMatrix = tf.concat(inputTensors, 1)
        # Run each card through model to get probability of picking each card
        predictions = self.model.predict(inputMatrix)
        pickedCardIndex = self.predictionPicker(predictions)
        pickedCard = uniqueCards[pickedCardIndex]
        # Save picked Card as index array
        self.pickedActions.append(inputMatrix.numpy()[pickedCardIndex])
        print({'damage':pickedCard.attack, 'cost':pickedCard.cost})
        self.turns += 1
        return [pickedCard]

class PickCardAtATimeDensePolicy(BaseRLPolicy):
    def __init__(self, encoder, predictionPicker, hiddenLayers=0, learningRate = 0.1):
        super().__init__(encoder, predictionPicker, learningRate)
        self.inputUnits = self.lenUniqueCards
        self.model = tf.keras.Sequential()
        self.model.add(tf.keras.layers.Dense(config.HIDDEN_UNITS, input_shape=(self.inputUnits,), use_bias=True))
        # self.model.add(tf.keras.layers.LayerNormalization(axis=-1))
        for i in range(hiddenLayers):
            self.model.add(tf.keras.layers.Dense(config.HIDDEN_UNITS, input_shape=(config.HIDDEN_UNITS,), use_bias=True))
        self.model.add(tf.keras.layers.Dense(1, use_bias=True))
        self.model.compile(loss='meanSquaredError', optimizer=tf.keras.optimizers.Adam(learningRate))
        self.encodedCardToColumnTensor = self.encodedCardToColumnTensor.bind(self)

    def update(self, score):
        avgScore = score / self.turns
        inputMatrix = tf.convert_to_tensor(self.pickedActions, dtype=tf.float32)
        labels = tf.convert_to_tensor([avgScore]*len(self.pickedActions), dtype=tf.float32)
        self.updateModel(inputMatrix, labels)

    def encodedCardToColumnTensor(self, encodedCard, inputUnits):
        combinationTensor = tf.reshape(tf.convert_to_tensor(encodedCard, dtype=tf.float32), [inputUnits, 1])
        return combinationTensor

    def pickCards(self, hand, world):
        # encode cards into tensors
        uniqueCards = self.getUniqueCards(hand)
        inputTensors = [self.encodedCardToColumnTensor(self.encoder.encodeCard(n, world), self.inputUnits) for n in uniqueCards]
        # Build matrix of permutations
        inputMatrix = tf.concat(inputTensors, 1)
        # Run each card through model to get probability of picking each card
        predictions = self.model.predict(inputMatrix)
        pickedCardIndex = self.predictionPicker(predictions)
        pickedCard = uniqueCards[pickedCardIndex]
        # Save picked Card as index array
        self.pickedActions.append(inputMatrix.numpy()[pickedCardIndex])
        print({'damage':pickedCard.attack, 'cost':pickedCard.cost})
        self.turns += 1
        return [pickedCard]


class HandPermutationDensePolicy(BaseRLPolicy):
    def __init__(self, encoder, predictionPicker, hiddenLayers=0, learningRate = 0.1):
        super().__init__(encoder, predictionPicker, learningRate)
        self.inputUnits = self.lenUniqueCards*config.HERO_BUDGET
        self.model = tf.keras.Sequential()
        self.model.add(tf.keras.layers.Dense(config.HIDDEN_UNITS, input_shape=(self.inputUnits,), use_bias=True))
        for i in range(hiddenLayers):
            self.model.add(tf.keras.layers.Dense(config.HIDDEN_UNITS, input_shape=(config.HIDDEN_UNITS,), use_bias=True))
        self.model.add(tf.keras.layers.Dense(1, use_bias=True))
        self.model.compile(loss='meanSquaredError', optimizer=tf.keras.optimizers.Adam(self.learningRate))
        self.oneHotCombinationToColumnTensor = self.oneHotCombinationToColumnTensor.bind(self)

    def update(self, score):
        avgScore = score / self.turns
        inputMatrix = tf.convert_to_tensor(self.pickedActions, dtype=tf.float32)
        labels = tf.convert_to_tensor([avgScore]*len(self.pickedActions), dtype=tf.float32)
        self.updateModel(inputMatrix, labels)

    def oneHotCombinationToColumnTensor(self, oneHotCombination, inputUnits):
        combinationTensor = tf.convert_to_tensor(oneHotCombination, dtype=tf.float32)
        paddedCombinationTensor = self.padCombinationTensor(combinationTensor)
        # reshape combinationTensort to be 1 column
        reshapedCombination = tf.reshape(paddedCombinationTensor, [inputUnits, 1])      
        return reshapedCombination

    def getProbabilityForPermutation(self, permutation, allPermutations, probabilities):
        for i in range(len(allPermutations)):
            if permutation == allPermutations[i]:
                return probabilities[i]
        raise Exception('Permutation not found')

    def pickCards(self, hand, world):
        # Make all combinations of up to n cards from hand
        possiblePlayedCards = self.buildNCardPermutations(config.HERO_BUDGET, hand)
        inputTensors = [self.oneHotCombinationToColumnTensor(self.encoder.oneHotEncodeCards(n), self.inputUnits) for n in possiblePlayedCards]
        # Build matrix of permutations
        inputMatrix = tf.concat(inputTensors, 1)
        # Run each combination through model to get probability of picking each card
        predictions = self.model.predict(inputMatrix)
        pickedCombinationIndex = self.predictionPicker(predictions)
        pickedCombination = possiblePlayedCards[pickedCombinationIndex]
        # var test = inputMatrix.arraySync()
        # Save actions for update
        self.pickedActions.append(inputMatrix[pickedCombinationIndex].numpy())
        print(pickedCombination)
        newHistory = [n.attack for n in pickedCombination]
        self.history.append(newHistory)
        self.turns += 1
        return pickedCombination





